### PR TITLE
Added matchers for cats.effect.IO

### PR DESCRIPTION
**Pull request by GW00D**
Added matchers for cats.effect.IO

### CODE REVIEW

File: cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+import scala.concurrent.TimeoutException
+import scala.concurrent.duration.{Duration, FiniteDuration}
+import org.specs2.text.NotNullStrings._
+import scala.concurrent.ExecutionContext.Implicits.global
+import cats.effect.Timer
+
+trait IOMatchers {
+
+  implicit val timer = Timer[IO]
```
Comment (etorreborre) [inline cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala]: Can you please add the return type?
```
implicit val timer: Timer[IO] = 
  Timer[IO]
```
This is recommended practice for implicit values (I think it is now mandatory in Dotty)

File: cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+  implicit def returnOk[T]: IOMatcher[T] =
```
Comment (etorreborre) [inline cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala]: Why do those matchers need to be implicit?

File: cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    attemptRun(ValueCheck.alwaysOk, None)
+
+  implicit def returnValue[T](check: ValueCheck[T]): IOMatcher[T] =
+    attemptRun(check, None)
+
+  implicit def returnBefore[T](duration: Duration): IOMatcher[T] =
```
Comment (etorreborre) [inline cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala]: I think we should better require a `FiniteDuration` here.

File: cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    private def errorResult[S <: IO[T]](e: Expectable[S])(t: Throwable): MatchResult[S] = {
+      val message = "an exception was thrown "+t.getMessage.notNull+" "+t.getClass.getName
+      result(false, message, message, e)
+    }
+
+    private def checkResult[S <: IO[T]](e: Expectable[S])(t: T): MatchResult[S] =
+      result(check.check(t), e)
+
+    def asFiniteDuration(d: Duration): FiniteDuration = Duration.fromNanos(d.toNanos)
```
Comment (etorreborre) [inline cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala]: Then we won't need that line.

File: guide/src/test/scala/org/specs2/guide/matchers/IOMatchers.scala [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+object IOMatchers extends UserGuideCard {
+  def title = "IO Matchers"
+  def text = s2"""
+  There are some basic matchers to check `cats.effect.IO` instances:
+
+ * `returnOk` to check if an `IO` completes successfully
+ * `returnBefore(duration)` to check if an `IO` completes successfully before the specified duration
+ * `returnValue(matcher)` to check if an `IO` successfully returns a value that satisfies the matcher
+ * `returnValue(function: A => AsResult[B])` to check if an `IO` completes successfully with a value `a`, where `function(a)` returns a successful [Result](org.specs.guide.AsResultTypeclass.html). This can be used to do conditional assertions that depend on which value is returned by your `IO`
```
Comment (etorreborre) [inline guide/src/test/scala/org/specs2/guide/matchers/IOMatchers.scala]: Is it possible to break this line to fit the screen?

File: cats/jvm/src/test/scala/org/specs2/matcher/IOMatchersSpec.scala [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+      { (IO(1) must returnValue(2)) returns "1 != 2" }
+    }
+
+    "check that the task finishes before a given time" >> {
+      { IO(1) must returnBefore(10.millis) }
+      { (IO { Thread.sleep(50); 1 } must returnBefore(10.millis)) returns "Timeout after 10 milliseconds" }
+    }
+
+    "check the return value with a timeout" >> {
+      { IO { Thread.sleep(5); 1 } must returnBefore(10.millis).withValue(1) }
```
Comment (etorreborre) [inline cats/jvm/src/test/scala/org/specs2/matcher/IOMatchersSpec.scala]: Nice!

File: cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala (hunk 1 above)
Comment (GW00D) [reply cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala]: Thanks :+1:

File: cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+import scala.concurrent.TimeoutException
+import scala.concurrent.duration.FiniteDuration
+import org.specs2.text.NotNullStrings._
+import scala.concurrent.ExecutionContext.Implicits.global
+import cats.effect.Timer
+
+trait IOMatchers {
+
+  val timer: Timer[IO] = Timer[IO]
```
Comment (etorreborre) [inline cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala]: I don't see where the `timer` is used, is it needed for anything?

File: cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+  def returnOk[T]: IOMatcher[T] =
+    attemptRun(ValueCheck.alwaysOk, None)
+
+  implicit def returnValue[T](check: ValueCheck[T]): IOMatcher[T] =
```
Comment (etorreborre) [inline cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala]: Is the `implicit` necessary here?

File: cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    attemptRun(check, None)
+
+  implicit def returnBefore[T](duration: FiniteDuration): IOMatcher[T] =
```
Comment (etorreborre) [inline cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala]: And there?

File: cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala (hunk 7 above)
Comment (GW00D) [reply cats/shared/src/main/scala/org/specs2/matcher/IOMatchers.scala]: Sorry - thanks for catching that!

**Comment by GW00D** [comment]
@etorreborre Updated again

**Comment by etorreborre** [comment]
Great thank you!

**Pull request merged**